src/execution/sql/runtime_types.cpp

```diff
@@ -655,4 +658,711 @@ Timestamp Timestamp::FromYMDHMSMU(int32_t year, int32_t month, int32_t day, int3
+
+  uint32_t magic_p = magic_map128_bit_constant_division[constant].p_ - 128;
+
+  if (magic_map128_bit_constant_division[constant].algo_ == 0) {
+    // Overflow Algorithm 1 - Magic number is < 2^128
+
+    uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+    value_ = result_upper >> magic_p;
+  } else {
+    // Overflow Algorithm 2 - Magic number is > 2^128
+
+    uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+    uint128_t add_upper = value_;
+
+    /*Perform addition*/
+    result_upper += add_upper;
+
+    auto carry = static_cast<uint128_t>(result_upper < add_upper);
+    carry = carry << 127;
+    // shrxi 1
+    result_upper = result_upper >> 1;
+    result_upper |= carry;
+
+    value_ = result_upper >> (magic_p - 1);
+  }
+}
+
+template <typename T>
+void Decimal<T>::SignedMultiplyWithDecimal(Decimal<T> input, uint32_t lower_precision) {
+  bool negative_result = (value_ < 0) != (input.GetValue() < 0);
+
+  // Not used Hacker Delight 2-14 because shift needs to be agnostic of underlying T
+  // Will be needed to change in the future when storage optimizations happen
+  if (value_ < 0) {
+    value_ = 0 - value_;
+  }
+
+  if (input.GetValue() < 0) {
+    input.SetValue(-input.GetValue());
+  }
+
+  MultiplyAndSet(input, lower_precision);
+
+  if (negative_result) {
+    value_ = 0 - value_;
+  }
+}
+
+template <typename T>
+void Decimal<T>::SignedMultiplyWithConstant(int64_t input) {
+  bool negative_result = (value_ < 0) != (input < 0);
+
+  // Not used Hacker Delight 2-14 because shift needs to be agnostic of underlying T
+  // Will be needed to change in the future when storage optimizations happen
+  if (value_ < 0) {
+    value_ = 0 - value_;
+  }
+
+  constexpr const uint128_t bottom_mask = (uint128_t{1} << 64) - 1;
+  constexpr const uint128_t top_mask = ~bottom_mask;
+
+  // First input
+  uint128_t a = value_;
+
+  // Second input
+  uint128_t b;
+  if (input < 0) {
+    b = input;
+  } else {
+    b = -input;
+  }
+
+  // Split into half words
+  uint128_t half_words_a[2];
+  uint128_t half_words_b[2];
+
+  half_words_a[0] = a & bottom_mask;
+  half_words_a[1] = (a & top_mask) >> 64;
+
+  half_words_b[0] = b & bottom_mask;
+  half_words_b[1] = (b & top_mask) >> 64;
+
+  // Calculate 256 bit result
+  uint128_t half_words_result[4];
+  CalculateMultiWordProduct128(half_words_a, half_words_b, half_words_result, 2, 2);
+
+  if (half_words_result[2] == 0 && half_words_result[3] == 0) {
+    value_ = half_words_result[0] | (half_words_result[1] << 64);
+  } else {
+    throw EXECUTION_EXCEPTION(fmt::format("Result overflow > 128 bits"), common::ErrorCode::ERRCODE_DATA_EXCEPTION);
+  }
+
+  if (negative_result) {
+    value_ = 0 - value_;
+  }
+}
+
+template <typename T>
+void Decimal<T>::SignedDivideWithConstant(int64_t input) {
+  bool negative_result = (value_ < 0) != (input < 0);
+
+  // Not used Hacker Delight 2-14 because shift needs to be agnostic of underlying T
```

Comment:
Not used???

Reply:
Also this comment is copy pasted everywhere, is this the right figure / section?

Reply:
Yes

Reply:
I am not stupid :P